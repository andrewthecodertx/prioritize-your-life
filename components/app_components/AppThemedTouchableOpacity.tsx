import React from "react";
import { ScaledSheet, s } from "react-native-size-matters";
import {
  StyleProp,
  TouchableOpacity,
  TouchableOpacityProps,
  ViewStyle,
} from "react-native";
import { AppThemedText } from "./AppThemedText";
import { COLORS } from "@/constants/Colors";

export type ThemedAppThemedTouchableOpacity = TouchableOpacityProps & {
  children: React.ReactNode;
  textStyles?: StyleProp<ViewStyle>;
  btnStyles?: StyleProp<ViewStyle>;
  onPress?: () => void;
};

/**
 * Custom TouchableOpacity component for the app.
 *
 * @component
 * @param {TouchableOpacityProps} props - The TouchableOpacity props.
 * @returns {JSX.Element} - The rendered TouchableOpacity component.
 */
const AppThemedTouchableOpacity = ({
  children,
  onPress,
  textStyles,
  btnStyles,
  disabled
}: ThemedAppThemedTouchableOpacity): JSX.Element => {
  return (
    <TouchableOpacity disabled={disabled} onPress={onPress} style={[styles.button, btnStyles]}>
      <AppThemedText style={[styles.buttonText, textStyles]}>
        {children}
      </AppThemedText>
    </TouchableOpacity>
  );
};

export default AppThemedTouchableOpacity;

const styles = ScaledSheet.create({
  button: {
    alignItems: "center",
    backgroundColor: COLORS.primary,
    borderRadius: s(5),
    marginBottom: s(10),
    padding: s(10),
    width: "80%",
  },
  buttonText: {
    color: COLORS.white,
  },
});
