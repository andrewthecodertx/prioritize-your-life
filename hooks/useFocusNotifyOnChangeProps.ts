import React from "react";
import { NotifyOnChangeProps } from "@tanstack/query-core";
import { useFocusEffect } from "@react-navigation/native";


/**
 * Hook that returns a function that can be used to notify the parent component
 * when the component is focused and the props have changed.
 *
 * This is useful when you want to notify the parent component when the props
 * have changed, but only when the component is focused.
 *
 * @param notifyOnChangeProps - The props to notify the parent component about
 * when they change. Can be a function that returns the props to notify about.
 */
export function useFocusNotifyOnChangeProps(
  notifyOnChangeProps?: NotifyOnChangeProps,
) {
  const focusedRef = React.useRef(true)

  useFocusEffect(
    React.useCallback(() => {
      focusedRef.current = true

      return () => {
        focusedRef.current = false
      }
    }, []),
  )

  return () => {
    if (!focusedRef.current) {
      return []
    }

    if (typeof notifyOnChangeProps === "function") {
      return notifyOnChangeProps()
    }

    return notifyOnChangeProps
  }
}
