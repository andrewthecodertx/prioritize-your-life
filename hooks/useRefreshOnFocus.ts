import React from "react";
import { useFocusEffect } from "@react-navigation/native";

/**
 * Refresh the screen when it is focused
 * @param refetch - The refetch function
 */
export function useRefreshOnFocus<T>(refetch: () => Promise<T>) {
  const firstTimeRef = React.useRef(true);

  useFocusEffect(
    React.useCallback(() => {
      if (firstTimeRef.current) {
        firstTimeRef.current = false;
        return;
      }

      refetch();
    }, [refetch])
  );
}
