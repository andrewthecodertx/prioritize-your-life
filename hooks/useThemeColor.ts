/**
 * Learn more about light and dark modes:
 * https://docs.expo.dev/guides/color-schemes/
 */

import { useColorScheme } from "react-native";
import { Colors } from "@/constants/Colors";

/**
 * Get the color for the current theme
 * @param props - The props object
 * @param colorName - The color name
 * @returns The color for the current theme
 */
export function useThemeColor(
  props: { light?: string; dark?: string },
  colorName: keyof typeof Colors.light & keyof typeof Colors.dark
) {
  const theme = useColorScheme() ?? "light";
  const colorFromProps = props[theme];

  if (colorFromProps) {
    return colorFromProps;
  } else {
    return Colors[theme][colorName];
  }
}
