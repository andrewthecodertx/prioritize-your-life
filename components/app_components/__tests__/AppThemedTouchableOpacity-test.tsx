import * as React from "react";
import renderer from "react-test-renderer";
import { TouchableOpacity } from "react-native";
import AppThemedTouchableOpacity from "../../app_components/AppThemedTouchableOpacity";
import { useThemeColor } from "@/hooks/useThemeColor";

describe("AppThemedTouchableOpacity Tests", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it(`renders correctly`, () => {
    const tree = renderer
      .create(
        <AppThemedTouchableOpacity>Children Content</AppThemedTouchableOpacity>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`calls onPress when pressed`, () => {
    const onPress = jest.fn();
    const component = renderer.create(
      <AppThemedTouchableOpacity onPress={onPress}>
        Children Content
      </AppThemedTouchableOpacity>
    );
    const touchableOpacity = component.root.findByType(TouchableOpacity);
    touchableOpacity.props.onPress();

    expect(onPress).toHaveBeenCalled();
  });

  it("renders correctly with dark theme", () => {
    const mockDarkColor = "#000000";
    (useThemeColor as jest.Mock).mockReturnValue(mockDarkColor);

    const tree = renderer
      .create(
        <AppThemedTouchableOpacity>Children Content</AppThemedTouchableOpacity>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with light theme", () => {
    const mockLightColor = "#ffffff";
    (useThemeColor as jest.Mock).mockReturnValue(mockLightColor);

    const tree = renderer
      .create(
        <AppThemedTouchableOpacity>Children Content</AppThemedTouchableOpacity>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
