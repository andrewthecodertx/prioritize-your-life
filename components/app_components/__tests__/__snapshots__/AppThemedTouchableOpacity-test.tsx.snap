// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppThemedTouchableOpacity Tests renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#CD03EF",
      "borderRadius": 10.714285714285714,
      "marginBottom": 21.428571428571427,
      "opacity": 1,
      "padding": 21.428571428571427,
      "width": "80%",
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#ECEDEE",
        },
        {
          "fontSize": 34.285714285714285,
          "lineHeight": 51.42857142857143,
        },
        undefined,
        undefined,
        undefined,
        undefined,
        [
          {
            "color": "#fff",
          },
          undefined,
        ],
      ]
    }
  >
    Children Content
  </Text>
</View>
`;

exports[`AppThemedTouchableOpacity Tests renders correctly with dark theme 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#CD03EF",
      "borderRadius": 10.714285714285714,
      "marginBottom": 21.428571428571427,
      "opacity": 1,
      "padding": 21.428571428571427,
      "width": "80%",
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#000000",
        },
        {
          "fontSize": 34.285714285714285,
          "lineHeight": 51.42857142857143,
        },
        undefined,
        undefined,
        undefined,
        undefined,
        [
          {
            "color": "#fff",
          },
          undefined,
        ],
      ]
    }
  >
    Children Content
  </Text>
</View>
`;

exports[`AppThemedTouchableOpacity Tests renders correctly with light theme 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#CD03EF",
      "borderRadius": 10.714285714285714,
      "marginBottom": 21.428571428571427,
      "opacity": 1,
      "padding": 21.428571428571427,
      "width": "80%",
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#ffffff",
        },
        {
          "fontSize": 34.285714285714285,
          "lineHeight": 51.42857142857143,
        },
        undefined,
        undefined,
        undefined,
        undefined,
        [
          {
            "color": "#fff",
          },
          undefined,
        ],
      ]
    }
  >
    Children Content
  </Text>
</View>
`;
